rules/tock.bzl

```diff
@@ -0,0 +1,127 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Rules for assembling Tock binaries.
+"""
+
+load("@rules_cc//cc:find_cc_toolchain.bzl", "find_cc_toolchain")
+load(
+    "//rules:rv.bzl",
+    "rv_rule",
+    _OPENTITAN_CPU = "OPENTITAN_CPU",
+    _OPENTITAN_PLATFORM = "OPENTITAN_PLATFORM",
+)
+
+TockApplication = provider(
+    fields = {
+        "tab": "TAB file for this application",
+        "tbf": "TBF file for this application",
+        "elf": "ELF file for this application",
+    },
+)
+
+def _tock_elf2tab_impl(ctx):
+    name = ctx.attr.package_name if ctx.attr.package_name else ctx.attr.name
+
+    tabfile = ctx.actions.declare_file("{}.tab".format(name))
+    tbffile = ctx.actions.declare_file("{}.tbf".format(name))
+
+    elffile = ctx.actions.declare_file("{}.elf".format(name))
+    ctx.actions.symlink(output = elffile, target_file = ctx.file.src)
+
+    outputs = [tabfile, tbffile]
+    args = [
+        "--kernel-major={}".format(ctx.attr.kernel_major),
+        "--kernel-minor={}".format(ctx.attr.kernel_minor),
+        "--package-name={}".format(name),
+        "--output-file={}".format(tabfile.path),
+    ]
+    if ctx.attr.protected_region_size:
+        args.append("--protected-region-size={}".format(ctx.attr.protected_region_size))
+    if ctx.attr.stack:
+        args.append("--stack={}".format(ctx.attr.stack))
+    if ctx.attr.verbose:
+        args.append("--verbose")
+    args.append(elffile.path)
+
+    ctx.actions.run(
+        mnemonic = "ELF2TAB",
+        outputs = outputs,
+        inputs = [elffile, ctx.executable._elf2tab],
+        arguments = args,
+        executable = ctx.executable._elf2tab,
+    )
+
+    outputs.append(elffile)
```

Comment:
Can this be inlined into the `outputs` above?

Reply:
It could, but the _proper_ order for things is positional arguments last :)